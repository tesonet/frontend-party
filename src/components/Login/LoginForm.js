import styled from 'styled-components';

export default styled.form`
  display: flex;
  flex-direction: column;
  margin: 0 auto;
  position: relative;
  @media (max-width: 400px) {
    width: 90%;
  }
`;
