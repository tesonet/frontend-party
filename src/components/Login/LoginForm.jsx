import React, {
    Fragment, useRef, memo,
} from 'react';
import PropTypes from 'prop-types';
import { useSelector } from 'react-redux';
import { Field, reduxForm } from 'redux-form';
import { Overlay, Tooltip } from 'react-bootstrap';
import LoginSubmitButton from './LoginSubmitButton';
import PasswordFieldWrapper from './Fields/PasswordFieldWrapper';
import UsernameFieldWrapper from './Fields/UsernameFieldWrapper';
import validate from '../../utils/validate';

const Form = (props) => {
    const { handleSubmit, error } = props;
    const submitted = useSelector((state) => state.auth.submitted);
    const formRef = useRef(null);

    return (
        <Fragment>
            <form
                onSubmit={handleSubmit}
                ref={formRef}
            >
                <Field
                    name="username"
                    component={UsernameFieldWrapper}
                    label="Username"
                />
                <Field
                    name="password"
                    component={PasswordFieldWrapper}
                    label="Password"
                />
                <LoginSubmitButton
                    disabled={!submitted}
                    className="login-button"
                />
            </form>
            <Overlay target={formRef.current} show={!!error} placement="bottom">
                {(tooltipProps) => (
                    <Tooltip id="button-tooltip" {...tooltipProps}>
                        {error}
                    </Tooltip>
                )}
            </Overlay>
        </Fragment>
    );
};

Form.defaultProps = {
    error: null,
};

Form.propTypes = {
    error: PropTypes.string,
    handleSubmit: PropTypes.func.isRequired,
};

const LoginForm = reduxForm({
    form: 'login',
    validate,
})(Form);

export default memo(LoginForm);
