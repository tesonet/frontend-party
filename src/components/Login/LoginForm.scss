@mixin commonTextProperties() {
  font-size     : 1rem;
  line-height   : 2rem;
  letter-spacing: 0.4px;
}

.loginForm {
  display       : flex;
  flex-direction: column;


  &__group {
    position: relative;

    &-input {
      height       : 3.6rem;
      width        : 100%;
      border       : none;
      border-radius: 0.3rem;
      padding      : 0 2rem 0 3.6rem;
      color        : #999;
      box-sizing   : border-box;
      @include commonTextProperties();

      &::placeholder {
        color: #B3B3B3;
      }
    }

    &-icon {
      position: absolute;
      left    : 1.6rem;
      top     : 1.2rem;
    }
  }

  &__error {
    padding-top: 0.2rem;
    font-size  : 0.8rem;
    min-height : 2rem;
    color      : lightpink;
  }

  &__button {
    height          : 3.6rem;
    width           : 100%;
    border          : none;
    border-radius   : 0.3rem;
    text-align      : center;
    background-color: #9FD533;
    transition      : background-color cubic-bezier(0.785, 0.135, 0.15, 0.86) 0.2s;
    cursor          : pointer;
    color           : white;
    font-weight     : 600;
    @include commonTextProperties();

    &:hover {
      background-color: #86b300;
    }
  }
}