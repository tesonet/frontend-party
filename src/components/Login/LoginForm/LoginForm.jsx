import React, { Component } from "react";
import axios from "axios";

// Config. REST API URL
import Config from "utils/Config";

// Components
import ValidateInput from "utils/ValidateInput";
import FormControl from "common/FormControl/FormControl";
import FormError from "common/FormError/FormError";
import Button from "common/Button/Button";

// Styles, images
import styles from "./LoginForm.module.scss";
import logo from "./images/testio.svg";

export default class Login extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isLoading: "",
      username: "",
      password: "",
      inputError: {},
      formError: ""
    };
    this.onSubmit = this.onSubmit.bind(this);
    this.onChange = this.onChange.bind(this);
  }

  onChange(e) {
    this.setState({ [e.target.name]: e.target.value });
  }

  isValid() {
    const { inputError, isValid } = ValidateInput(this.state);
    if (!isValid) {
      this.setState({ inputError });
    }
    return isValid;
  }

  onSubmit(e) {
    e.preventDefault();

    // Local state
    const { username, password } = this.state;

    if (this.isValid()) {
      // Reset error messages, and show loading
      this.setState({ inputError: {}, isLoading: "loading" });
      axios
        .post(Config.REST_API_URL + "tokens", {
          username,
          password
        })
        .then(res => {
          localStorage.setItem("testioToken", res.data.token);
          this.props.history.push(Config.DASHBOARD_URL);
        })
        .catch(err => {
          let msg;
          switch (err.response.status) {
            case 401:
              msg = "Wrong user name or password";
              break;
            case 503:
              msg = "Service unavailable";
              break;
            default:
              msg = "Something wrong";
          }
          this.setState({
            formError: msg,
            isLoading: ""
          });
        });
    }
  }

  render() {
    const { inputError, username, password, isLoading } = this.state;
    return (
      <div className={styles.login}>
        <img src={logo} alt="Testio" width="245" />
        <form onSubmit={this.onSubmit} className={styles.form}>
          {this.state.formError && <FormError text={this.state.formError} />}
          <FormControl
            cssModifier="form__control--user"
            type="text"
            placeholder="Username"
            name="username"
            value={username}
            onChange={this.onChange}
            error={inputError.username}
          />
          <FormControl
            cssModifier="form__control--pass"
            type="password"
            placeholder="Password"
            name="password"
            value={password}
            onChange={this.onChange}
            error={inputError.password}
          />

          <Button title="Login" loadingClass={isLoading} />
        </form>
        <footer>
          <a href="#">Create an Account</a> | <a href="#">Reset password</a>
        </footer>
      </div>
    );
  }
}
