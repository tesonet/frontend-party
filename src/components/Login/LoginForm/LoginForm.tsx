import React from 'react'
import { useFormik, FormikErrors } from 'formik'
import {
  Form,
  StyledInput,
  Button,
  ErrorContainer,
  IconContainer,
  InputWrapper,
} from './LoginForm.style'
import { Credentials } from 'types/auth'
import UserIcon from 'components/Icons/UserIcon'
import LockIcon from 'components/Icons/LockIcon'
import { AnyAction, Dispatch } from 'redux'

interface LoginFormProps {
  loading: boolean
  loadUser: (username: string, password: string) => (dispatch: Dispatch<AnyAction>) => void
}

interface InputProps {
  name: string
  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void
  value: string
  placeholder: string
  error: string | undefined | false
  type: string
  icon?: JSX.Element
  id: string
  onBlur: (e: React.FocusEvent<any>) => void
}

const Input: React.FC<InputProps> = ({
  name,
  id,
  onChange,
  value,
  placeholder,
  error,
  type,
  icon,
  onBlur,
}) => {
  return (
    <>
      <InputWrapper>
        <StyledInput
          id={id}
          name={name}
          type={type}
          onChange={onChange}
          value={value}
          placeholder={placeholder}
          error={!!error}
          autoComplete="new-password"
          onBlur={onBlur}
        />
        <IconContainer>{icon}</IconContainer>
      </InputWrapper>
      <ErrorContainer error={!!error}>{error}</ErrorContainer>
    </>
  )
}

const LoginForm: React.FC<LoginFormProps> = ({ loading, loadUser }) => {
  const formik = useFormik({
    initialValues: {
      login: '',
      password: '',
    },

    validate: (values: Credentials) => {
      const errors: FormikErrors<Credentials> = {}
      if (!values.login) {
        errors.login = 'Please enter your username'
      }
      if (!values.password) {
        errors.password = 'Please enter your password'
      }

      return errors
    },

    onSubmit: values => {
      loadUser(values.login, values.password)
      formik.resetForm()
    },
  })

  return (
    <Form onSubmit={formik.handleSubmit} autoComplete="off">
      <Input
        id="login"
        name="login"
        type="text"
        onChange={formik.handleChange}
        value={formik.values.login}
        placeholder="Username"
        error={formik.touched.login && formik.errors.login}
        icon={<UserIcon />}
        onBlur={formik.handleBlur}
      />

      <Input
        id="password"
        name="password"
        type="password"
        onChange={formik.handleChange}
        value={formik.values.password}
        placeholder="Password"
        error={formik.touched.password && formik.errors.password}
        icon={<LockIcon />}
        onBlur={formik.handleBlur}
      />

      <Button type="submit" disabled={loading || !!formik.errors.login || !!formik.errors.password}>
        Login
      </Button>
    </Form>
  )
}

export default LoginForm
