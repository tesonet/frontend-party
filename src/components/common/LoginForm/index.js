import React, {useState, useRef, createRef} from 'react'
import {useDispatch, useSelector} from 'react-redux'
import {loginUser} from '../../../modules/auth/actions'
import {getIsLoggingIn, getErrorMessage} from '../../../modules/auth/selectors'
import {createValidator, required, minLength} from '../../../utils/validation'
import Input from '../../../ui/Input'
import Button from '../../../ui/Button'
import Alert from '../../../ui/Alert'

const config = [
  {
    type: 'text',
    name: 'username',
    label: 'Username',
    icon: 'user',
  },
  {
    type: 'password',
    name: 'password',
    label: 'Password',
    icon: 'lock',
  },
]

const validate = createValidator({
  username: [required, minLength(3)],
  password: required,
})

// TODO: reuse hooks logic

export default () => {
  const dispatch = useDispatch()
  const isLoggingIn = useSelector(getIsLoggingIn)
  const errorMessage = useSelector(getErrorMessage)

  const inputsRef = useRef(config.map(() => createRef()))

  const [inputs, setInputState] = useState({
    username: '',
    password: '',
  })

  const [errors, setErrors] = useState({})

  const handleInputChange = event => {
    const {target} = event
    const {type, name} = target
    const value = type === 'checkbox' ? target.checked : target.value

    setInputState(prevState => {
      return {
        ...prevState,
        [name]: value,
      }
    })
  }

  const handleFormSubmit = event => {
    const inputErrors = validate(inputs) || {}
    const invalidInputNames = Object.keys(inputErrors)

    if (invalidInputNames.length) {
      setErrors(inputErrors)
      // focus 1st invalid field
      try {
        const invalidInputName = invalidInputNames[0]

        inputsRef.current.forEach(ref => {
          if (ref.current.name === invalidInputName) {
            ref.current.focus()
          }
        })
      } catch (error) {
        console.error(error)
      }
    } else {
      setErrors({})
      dispatch(loginUser(inputs))
    }

    event.preventDefault()
  }

  return (
    <form onSubmit={handleFormSubmit}>
      {config.map((input, index) => {
        const {name} = input
        return (
          <Input
            key={name}
            ref={inputsRef.current[index]}
            type={input.type}
            name={input.name}
            label={input.label}
            value={inputs[name]}
            onChange={handleInputChange}
            startIcon={input.icon}
            disabled={isLoggingIn}
            error={!!errors[name]}
            helperText={errors[name]}
            mb={3}
          />
        )
      })}
      {!!errorMessage && (
        <Alert data-test-id="login-error" variant="danger" mb={3}>
          {errorMessage}
        </Alert>
      )}
      <Button
        variant="contained"
        size="medium"
        type="submit"
        loading={isLoggingIn}
        fullWidth>
        Log In
      </Button>
    </form>
  )
}
