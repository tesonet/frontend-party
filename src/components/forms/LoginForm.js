import React from 'react';
import {Form,Button,Input,Message} from 'semantic-ui-react';
import Validator from 'validator';
import PropTypes from 'prop-types';
import ErrorMessage from '../messages/ErrorMessage';




class LoginForm extends React.Component{

    state = {
        data : {
            email : '',
            password : ''
        },
        loading : false,
        errors : {}
    }

    onChange = e => this.setState({
        data : {...this.state.data, [e.target.name] : e.target.value }
    });

    onSubmit = () => {
        const errors = this.validate(this.state.data);
        this.setState({
            errors
        });
        if (Object.keys(errors).length === 0){
            this.setState({ loading : true});
            this.props
            .submit(this.state.data)
            .catch(err => this.setState({
                errors : err.response.data.errors, loading: false
            }));
            
        } 
    };

    validate = data => {
        const errors = {};
        if (!Validator.isEmail(data.email)) errors.email = "Invalid Email"
        if (!data.password) errors.password = "This can not be empty";
        return errors;
    }

    render(){

        let ButtonStyle = {
            width: "360px",
            height: "56px",
            borderRadius: "5px",
            backgroundColor: "#9fd533",
            color : "#ffff",
        }  
        
        const { data, errors, loading } = this.state;

        return(
            <Form onSubmit={this.onSubmit} loading= {loading}>   
                {errors.global && (
                    <Message negative>
                <Message.Header>Something went wrong :(</Message.Header>
                <p>{errors.global}</p>
                </Message>  
                

            )}
                <Form.Field error = {!!errors.email} >
                    <Input
                        icon='user' 
                        iconPosition='left'
                        type = "email"
                        id = "email"
                        name = "email"
                        value = {data.email}
                        placeholder = "Username"
                        onChange = {this.onChange}
                    />
                    {errors.email && <ErrorMessage text = { errors.email } />}
                </Form.Field>

                <Form.Field error = {!!errors.password}>
                    <Input
                        icon='lock' 
                        iconPosition='left'
                        type = "password"
                        id = "password"
                        name = "password"
                        value = {data.password}
                        placeholder = "Password"
                        onChange = {this.onChange}
                    />
                    {errors.password && <ErrorMessage text = { errors.password } />}
                </Form.Field>
                 <Button style = {ButtonStyle}>Login</Button>
            </Form>
        );
    }
}

LoginForm.propTypes = {
    submit : PropTypes.func.isRequired
}

export default LoginForm;