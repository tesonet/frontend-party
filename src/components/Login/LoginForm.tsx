import React from 'react';
import { Form, Field, FormikTouched, FormikErrors } from 'formik';
import { Credentials } from './Login';

import './LoginForm.scss';

interface Props {
  touched: FormikTouched<Credentials>;
  errors: FormikErrors<Credentials>;
  isSubmitting: boolean;
}

const LoginForm: React.FC<Props> = ({ touched, errors, isSubmitting }) => (
  <Form className="loginForm">
    <div className="loginForm__group">
      <Field
        type="username"
        name="username"
        placeholder="Username"
        className="loginForm__group-input"
      />
      <img className="loginForm__group-icon" src="/static/images/person-icon.svg" />
      <div className="loginForm__error">{touched.username && errors.username || ''}</div>
    </div>

    <div className="loginForm__group">
      <Field
        type="password"
        name="password"
        placeholder="Password"
        className="loginForm__group-input"
      />
      <img className="loginForm__group-icon" src="/static/images/lock-icon.svg" />
      <div className="loginForm__error">{touched.password && errors.password || ''}</div>
    </div>

    <button
      type="submit"
      name="submit"
      className="loginForm__button"
    >
      Log In
    </button>
  </Form>
);

export default LoginForm;