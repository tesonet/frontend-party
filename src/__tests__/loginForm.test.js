import React from "react";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";

//Component.
import LoginForm from "../components/LoginForm/LoginForm";

const mockStore = configureMockStore();
const store = mockStore({});

it("renders correctly", () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <LoginForm />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
