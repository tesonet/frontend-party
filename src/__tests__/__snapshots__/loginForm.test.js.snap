// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="LoginForm__wrapper"
>
  <img
    alt="testio"
    className="ImageProxy LoginForm__figure"
    src="logo.svg"
  />
  <img
    alt="loader"
    className="ImageProxy LoginForm__loader"
    src="loader.svg"
  />
  
  <form
    className="LoginForm"
    onSubmit={[Function]}
  >
    <div
      className="LoginForm__input-username"
    >
      <input
        onChange={[Function]}
        placeholder="Username"
        type="text"
        value=""
      />
    </div>
    <div
      className="LoginForm__input-password"
    >
      <input
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value=""
      />
    </div>
    <input
      type="submit"
      value="Log In"
    />
  </form>
</div>
`;
